server/channels/web/handlers.go

```diff
@@ -429,6 +400,55 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func (h Handler) handleContextError(c *Context, w http.ResponseWriter, r *http.Request) {
+	if c.Err == nil {
+		return
+	}
+
+	// We're handling payload limit error here because it needs to be handled globally.
+	var maxBytesErr *http.MaxBytesError
+	// check if error is a MaxBytesError error, which occurs when you read more bytes from buffer than configured
+	if ok := errors.As(c.Err, &maxBytesErr); ok {
+		// replace the context error with this error if so,
+		newErr := model.NewAppError(c.Err.Where, "api.context.request_body_too_large.app_error", nil, "Use the setting `MaximumPayloadSizeBytes` in Mattermost config to configure allowed payload limit. Learn more about this setting in Mattermost docs at https://docs.mattermost.com/configure/environment-configuration-settings.html#maximum-payload-size", http.StatusRequestEntityTooLarge)
+		c.Err = newErr
+	}
```

Comment:
Can we add a test for this? Also, could you point me out to where `c.Err` is actually storing the error returned from `MaxBytesReader`?

Reply:
Done.

> Also, could you point me out to where c.Err is actually storing the error returned from MaxBytesReader?

Take this code for example - https://github.com/mattermost/mattermost/blob/27dfeddfff20479a1d1f8517ad42cb5ad102578e/server/channels/api4/channel_bookmark.go#L44

```go
var channelBookmark *model.ChannelBookmark
	err := json.NewDecoder(r.Body).Decode(&channelBookmark)
	if err != nil || channelBookmark == nil {
		c.SetInvalidParamWithErr("channelBookmark", err)
		return
	}
```

So whenever we unmarshal the request body in any api function, if that generates an error, we call `SetInvalidParamWithErr` that sets c.Err with an invalid param error and also wraps the original error, which, in the case of request body being larger than the configured limit, is an `http.MaxBytesError` instance. This is the   consistently behaviour across all API methods, both new and old.